plugins/rich-editor/src/scripts/quill/HistoryModule.ts

```diff
@@ -32,52 +31,107 @@ export default class HistoryModule extends BaseHistoryModule {
+        return hasDeleteRecordOfLoadingEmbed || hasAddRecordOfLoadingEmbed;
+    }
+
+    /**
+     * Does the operation contain a code-block insert?
+     * @param operation Content operation.
+     */
+    private isCodeBlock = (operation: DeltaOperation | undefined): boolean => {
+        return (
+            typeof operation === "object" &&
+            "attributes" in operation &&
+            typeof operation.attributes === "object" &&
+            CodeBlock.blotName in operation.attributes &&
+            operation.attributes[CodeBlock.blotName]
+        );
+    };
+
+    /**
+     * Does the operation include an external embed insert?
+     * @param operation Content operation.
+     */
+    private isEmbedInsert = (operation: DeltaOperation | undefined): boolean => {
+        return this.isInsertOperation(operation) && ExternalEmbedBlot.blotName in operation!.insert;
+    };
+
+    /**
+     * Does the operation include a fully-loaded external embed insert?
+     * @param operation Content operation.
+     */
+    private isEmbedLoadedInsert = (operation: DeltaOperation | undefined): boolean => {
+        return this.isEmbedInsert(operation) && "data" in operation!.insert[ExternalEmbedBlot.blotName];
+    };
+
+    /**
+     * Does the operation contain any inserts?
+     * @param operation Content operation.
+     */
+    private isInsertOperation = (operation: DeltaOperation | undefined): boolean => {
+        return typeof operation === "object" && "insert" in operation && typeof operation.insert === "object";
+    };
+
+    /**
+     * Iterate through a list of operations and, using a callback, determine if any contain a match.
+     * @param ops List of change operations.
+     * @param callback Function to check for the presence of a specific type of operation.
+     */
+    private operationsContain(ops: DeltaOperation[] | undefined, callback: (DeltaOperation) => boolean) {
+        if (!ops) {
+            return false;
+        }
+
+        for (const op of ops) {
+            if (callback(op)) {
+                return true;
             }
         }
-    };
 
+        return false;
+    }
+
+    /**
+     * Record changes made to the content.
+     * @param changeDelta New operations on the content.
+     * @param oldDelta Previous content state.
+     */
     public record(changeDelta: DeltaStatic, oldDelta: DeltaStatic) {
-        if (this.operationsContainKey(changeDelta.ops, this.EMBED_KEY)) {
+        if (this.operationsContain(changeDelta.ops, this.isEmbedInsert)) {
+            // Avoid merging several changes as single undo/redo if an embed was added.
             this.cutoff();
         }
 
-        if (this.operationsContainKey(changeDelta.ops, this.CODE_KEY)) {
+        if (this.operationsContain(changeDelta.ops, this.isCodeBlock)) {
+            // Avoid merging several changes as single undo/redo if a code block was added.
```

Comment:
Isn't the intention of `cutoff` to stop merging the changes into a single undo/redo action? For example, if I type "Hello world." and then create a code block beneath it, the cut off would stop "ld." and the code block from being undone as one transaction.

Maybe it would be clearer if the comment were something like:

```suggestion
            // Avoid merging preceding changes as single undo/redo when a code block is added.
```

Reply:
Oops yeah my bad. Something did feel off though.